File: backend-python/examples/android-kotlin/APIService.kt
// APIService.kt
// Example implementation for migrating to v2 API
//
// Key changes from v1:
// - Update BASE_URL to include /v2/
// - No other changes required!

package com.liftlog.api

import com.google.gson.annotations.SerializedName
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.UUID
import java.util.concurrent.TimeUnit

// MARK: - Configuration

object APIConfig {
    const val PRODUCTION_URL = "https://api.liftlog.com/v2/"
    const val STAGING_URL = "https://staging.liftlog.com/v2/"
    const val DEVELOPMENT_URL = "http://10.0.2.2:8000/v2/"  // Android emulator

    const val WEBSOCKET_PRODUCTION = "wss://api.liftlog.com/ai-chat"
    const val WEBSOCKET_STAGING = "wss://staging.liftlog.com/ai-chat"
    const val WEBSOCKET_DEVELOPMENT = "ws://10.0.2.2:8000/ai-chat"

    fun getBaseUrl(environment: Environment): String {
        return when (environment) {
            Environment.PRODUCTION -> PRODUCTION_URL
            Environment.STAGING -> STAGING_URL
            Environment.DEVELOPMENT -> DEVELOPMENT_URL
        }
    }
}

enum class Environment {
    PRODUCTION,
    STAGING,
    DEVELOPMENT
}

enum class AppStore(val value: String) {
    WEB("Web"),
    GOOGLE("Google"),
    APPLE("Apple"),
    REVENUECAT("RevenueCat")
}

// MARK: - Models (No changes from v1)

data class CreateUserResponse(
    val id: UUID,
    val lookup: String,
    val password: String
)

data class GetUserResponse(
    val id: UUID,
    val lookup: String,
    @SerializedName("encrypted_current_plan")
    val encryptedCurrentPlan: ByteArray?,
    @SerializedName("encrypted_profile_picture")
    val encryptedProfilePicture: ByteArray?,
    @SerializedName("encrypted_name")
    val encryptedName: ByteArray?,
    @SerializedName("encryption_iv")
    val encryptionIV: ByteArray,
    @SerializedName("rsa_public_key")
    val rsaPublicKey: ByteArray
)

data class PutUserDataRequest(
    val id: UUID,
    val password: String,
    @SerializedName("encrypted_current_plan")
    val encryptedCurrentPlan: ByteArray?,
    @SerializedName("encrypted_profile_picture")
    val encryptedProfilePicture: ByteArray?,
    @SerializedName("encrypted_name")
    val encryptedName: ByteArray?,
    @SerializedName("encryption_iv")
    val encryptionIV: ByteArray,
    @SerializedName("rsa_public_key")
    val rsaPublicKey: ByteArray
)

data class GenerateAIWorkoutRequest(
    val attributes: AIWorkoutAttributes
)

data class AIWorkoutAttributes(
    val age: Int,
    val gender: String,
    @SerializedName("weight_kg")
    val weightKg: Double,
    @SerializedName("experience_level")
    val experienceLevel: String,
    val goals: String,
    @SerializedName("days_per_week")
    val daysPerWeek: Int,
    @SerializedName("additional_preferences")
    val additionalPreferences: String? = null
)

data class ProgramBlueprint(
    val name: String,
    val description: String,
    @SerializedName("duration_weeks")
    val durationWeeks: Int,
    val sessions: List<SessionBlueprint>
)

data class SessionBlueprint(
    val name: String,
    @SerializedName("estimated_duration_minutes")
    val estimatedDurationMinutes: Int,
    val exercises: List<ExerciseBlueprint>
)

data class ExerciseBlueprint(
    val name: String,
    val sets: Int,
    @SerializedName("reps_min")
    val repsMin: Int,
    @SerializedName("reps_max")
    val repsMax: Int,
    @SerializedName("rest_seconds")
    val restSeconds: Int,
    val notes: String? = null
)

// MARK: - API Interface

interface LiftLogAPI {

    // User Management
    @POST("user/create")
    suspend fun createUser(@Body request: Map<String, Any> = emptyMap()): CreateUserResponse

    @GET("user/{idOrLookup}")
    suspend fun getUser(@Path("idOrLookup") idOrLookup: String): GetUserResponse

    @PUT("user")
    suspend fun updateUser(@Body request: PutUserDataRequest)

    @POST("user/delete")
    suspend fun deleteUser(@Body request: Map<String, Any>)

    // AI Endpoints (require Authorization header)
    @POST("ai/workout")
    suspend fun generateAIWorkout(
        @Header("Authorization") authorization: String,
        @Body request: GenerateAIWorkoutRequest
    ): ProgramBlueprint

    @POST("ai/session")
    suspend fun generateAISession(
        @Header("Authorization") authorization: String,
        @Body request: GenerateAIWorkoutRequest
    ): SessionBlueprint
}

// MARK: - Interceptors

class AuthorizationInterceptor(
    private val appStore: AppStore,
    private val getToken: () -> String
) : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val original = chain.request()

        // Only add auth to AI endpoints
        if (original.url.encodedPath.contains("/ai/")) {
            val token = getToken()
            val request = original.newBuilder()
                .header("Authorization", "Bearer ${appStore.value} $token")
                .build()
            return chain.proceed(request)
        }

        return chain.proceed(original)
    }
}

class RateLimitInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val response = chain.proceed(chain.request())

        // Handle rate limiting
        if (response.code == 429) {
            val retryAfter = response.header("Retry-After")
            // Log or handle rate limit
            println("Rate limited. Retry after: $retryAfter")
        }

        return response
    }
}

// MARK: - API Service

class APIService private constructor(
    private val environment: Environment
) {

    private val api: LiftLogAPI

    init {
        val client = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .addInterceptor(RateLimitInterceptor())
            .build()

        val retrofit = Retrofit.Builder()
            .baseUrl(APIConfig.getBaseUrl(environment))
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        api = retrofit.create(LiftLogAPI::class.java)
    }

    // Singleton instances
    companion object {
        @Volatile
        private var INSTANCE: APIService? = null

        fun getInstance(environment: Environment = Environment.PRODUCTION): APIService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: APIService(environment).also { INSTANCE = it }
            }
        }
    }

    // User endpoints
    suspend fun createUser(): CreateUserResponse {
        return api.createUser()
    }

    suspend fun getUser(idOrLookup: String): GetUserResponse {
        return api.getUser(idOrLookup)
    }

    suspend fun updateUser(request: PutUserDataRequest) {
        api.updateUser(request)
    }

    // AI endpoints
    suspend fun generateWorkout(
        attributes: AIWorkoutAttributes,
        appStore: AppStore,
        purchaseToken: String
    ): ProgramBlueprint {
        val auth = "Bearer ${appStore.value} $purchaseToken"
        return api.generateAIWorkout(auth, GenerateAIWorkoutRequest(attributes))
    }
}

// MARK: - Usage Example

/*
// In your Activity or ViewModel:

class WorkoutViewModel : ViewModel() {
    private val apiService = APIService.getInstance(Environment.PRODUCTION)

    fun loadUser(userId: String) {
        viewModelScope.launch {
            try {
                val user = apiService.getUser(userId)
                // Handle user data
                println("User: ${user.lookup}")
            } catch (e: Exception) {
                // Handle error
                println("Error: ${e.message}")
            }
        }
    }

    fun generateWorkout() {
        viewModelScope.launch {
            val attributes = AIWorkoutAttributes(
                age = 30,
                gender = "male",
                weightKg = 75.0,
                experienceLevel = "Intermediate",
                goals = "Build muscle",
                daysPerWeek = 4
            )

            try {
                val program = apiService.generateWorkout(
                    attributes = attributes,
                    appStore = AppStore.GOOGLE,
                    purchaseToken = "your.purchase.token.here"
                )
                // Handle program
                println("Program: ${program.name}")
            } catch (e: Exception) {
                // Handle error (including rate limiting)
                println("Error: ${e.message}")
            }
        }
    }
}
*/

File: backend-python/examples/ios-swift/APIService.swift
/// APIService.swift
/// Example implementation for migrating to v2 API
///
/// Key changes from v1:
/// - Add /v2 to base URL
/// - No other changes required!

import Foundation

// MARK: - Configuration

enum APIEnvironment {
    case production
    case staging
    case development

    var baseURL: String {
        switch self {
        case .production:
            return "https://api.liftlog.com/v2"
        case .staging:
            return "https://staging.liftlog.com/v2"
        case .development:
            return "http://localhost:8000/v2"
        }
    }

    var websocketURL: String {
        switch self {
        case .production:
            return "wss://api.liftlog.com/ai-chat"
        case .staging:
            return "wss://staging.liftlog.com/ai-chat"
        case .development:
            return "ws://localhost:8000/ai-chat"
        }
    }
}

// MARK: - Models (No changes from v1)

struct CreateUserResponse: Codable {
    let id: UUID
    let lookup: String
    let password: String
}

struct GetUserResponse: Codable {
    let id: UUID
    let lookup: String
    let encryptedCurrentPlan: Data?
    let encryptedProfilePicture: Data?
    let encryptedName: Data?
    let encryptionIV: Data
    let rsaPublicKey: Data

    enum CodingKeys: String, CodingKey {
        case id, lookup
        case encryptedCurrentPlan = "encrypted_current_plan"
        case encryptedProfilePicture = "encrypted_profile_picture"
        case encryptedName = "encrypted_name"
        case encryptionIV = "encryption_iv"
        case rsaPublicKey = "rsa_public_key"
    }
}

struct PutUserDataRequest: Codable {
    let id: UUID
    let password: String
    let encryptedCurrentPlan: Data?
    let encryptedProfilePicture: Data?
    let encryptedName: Data?
    let encryptionIV: Data
    let rsaPublicKey: Data

    enum CodingKeys: String, CodingKey {
        case id, password
        case encryptedCurrentPlan = "encrypted_current_plan"
        case encryptedProfilePicture = "encrypted_profile_picture"
        case encryptedName = "encrypted_name"
        case encryptionIV = "encryption_iv"
        case rsaPublicKey = "rsa_public_key"
    }
}

// MARK: - API Service

class APIService {
    static let shared = APIService()

    private let environment: APIEnvironment
    private let session: URLSession

    init(environment: APIEnvironment = .production) {
        self.environment = environment

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 300
        self.session = URLSession(configuration: config)
    }

    var baseURL: String {
        environment.baseURL
    }

    // MARK: - User Endpoints

    /// Create a new user
    /// POST /v2/user/create
    func createUser() async throws -> CreateUserResponse {
        let url = URL(string: "\(baseURL)/user/create")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = "{}".data(using: .utf8)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.httpError(httpResponse.statusCode)
        }

        return try JSONDecoder().decode(CreateUserResponse.self, from: data)
    }

    /// Get user by ID or lookup
    /// GET /v2/user/{idOrLookup}
    func getUser(idOrLookup: String) async throws -> GetUserResponse {
        let url = URL(string: "\(baseURL)/user/\(idOrLookup)")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.httpError(httpResponse.statusCode)
        }

        return try JSONDecoder().decode(GetUserResponse.self, from: data)
    }

    /// Update user data
    /// PUT /v2/user
    func updateUser(request: PutUserDataRequest) async throws {
        let url = URL(string: "\(baseURL)/user")!
        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "PUT"
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let encoder = JSONEncoder()
        urlRequest.httpBody = try encoder.encode(request)

        let (_, response) = try await session.data(for: urlRequest)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.httpError(httpResponse.statusCode)
        }
    }

    // MARK: - AI Endpoints (require purchase token)

    /// Generate AI workout plan
    /// POST /v2/ai/workout
    /// Requires: Authorization: Bearer {AppStore} {ProToken}
    func generateAIWorkout(
        attributes: AIWorkoutAttributes,
        appStore: AppStore,
        purchaseToken: String
    ) async throws -> ProgramBlueprint {
        let url = URL(string: "\(baseURL)/ai/workout")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(appStore.rawValue) \(purchaseToken)",
                        forHTTPHeaderField: "Authorization")

        let body = ["attributes": attributes]
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        // Handle rate limiting
        if httpResponse.statusCode == 429 {
            if let retryAfter = httpResponse.value(forHTTPHeaderField: "Retry-After") {
                throw APIError.rateLimited(retryAfter: retryAfter)
            }
            throw APIError.rateLimited(retryAfter: nil)
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.httpError(httpResponse.statusCode)
        }

        return try JSONDecoder().decode(ProgramBlueprint.self, from: data)
    }
}

// MARK: - Supporting Types

enum AppStore: String {
    case web = "Web"
    case google = "Google"
    case apple = "Apple"
    case revenueCat = "RevenueCat"
}

enum APIError: Error {
    case invalidResponse
    case httpError(Int)
    case rateLimited(retryAfter: String?)
}

// Placeholder types
struct AIWorkoutAttributes: Codable {
    let age: Int
    let gender: String
    let weightKg: Double
    let experienceLevel: String
    let goals: String
    let daysPerWeek: Int

    enum CodingKeys: String, CodingKey {
        case age, gender, goals
        case weightKg = "weight_kg"
        case experienceLevel = "experience_level"
        case daysPerWeek = "days_per_week"
    }
}

struct ProgramBlueprint: Codable {
    let name: String
    let description: String
    let durationWeeks: Int
    let sessions: [SessionBlueprint]

    enum CodingKeys: String, CodingKey {
        case name, description, sessions
        case durationWeeks = "duration_weeks"
    }
}

struct SessionBlueprint: Codable {
    let name: String
    let estimatedDurationMinutes: Int
    let exercises: [ExerciseBlueprint]

    enum CodingKeys: String, CodingKey {
        case name, exercises
        case estimatedDurationMinutes = "estimated_duration_minutes"
    }
}

struct ExerciseBlueprint: Codable {
    let name: String
    let sets: Int
    let repsMin: Int
    let repsMax: Int
    let restSeconds: Int
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case name, sets, notes
        case repsMin = "reps_min"
        case repsMax = "reps_max"
        case restSeconds = "rest_seconds"
    }
}
